src/PointwiseFunctions/GeneralRelativity/KerrSchildCoords.hpp

```diff
@@ -0,0 +1,115 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <limits>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Options/Options.hpp"
+
+// IWYU pragma:  no_include <pup.h>
+
+/// \cond
+namespace PUP {
+class er;  // IWYU pragma: keep
+}  // namespace PUP
+/// \endcond
+
+namespace gr {
+/*!
+ * \brief Class containing functions related to transforming tensors between
+ * spherical and Cartesian Kerr-Schild coordinates in Kerr spacetime.
+ *
+ * The Cartesian Kerr-Schild coordinates \f$(x, y, z)\f$ are defined in terms of
+ * the spherical coordinates \f$(r, \theta, \phi)\f$ through
+ *
```

Comment:
It should be made clear that the `r` coordinate is the Boyer-Lindquist coordinate, and therefore that the Kerr horizon is at a constant value of `r`.  Perhaps the `r` here should be called `r_{BL}`?

I worry about confusion versus the usual spherical coordinates `r=x^2+y^2+z^2`, which one could also define for Kerr spacetime, and which we do indeed define (e.g. wave extraction surfaces of constant `r` and apparent-horizon radii `r` refer to the "usual" `r`, not the Boyer-Lindquist `r`)

Reply:
As you say, r, and also theta, are in fact the BL r and theta. Phi is not. I'll clarify the relations in the documentation. 

The name "spherical" was taken from what is used in the GRMHD community, as these functions are primarily used as helper functions in hydro solutions. I don't know of any better name, but if you have a good suggestion in order to distinguish it from the usual spherical coords, I'll be happy to adopt it. 

I should also mention that there will be a following commit (in a different PR) adding the Kerr metric in this coordinate system, which, again, will be used primarily to simplify some hydro solutions. "Spherical Kerr-Schild coordinates" is, again, the name people adopted for the metric in these coordinates. Feel free to suggest a better name.

Reply:
I'm still confused about theta and phi.
If x,y,z are the standard Kerr-Schild coordinates like in MTW exercise 33.8, then is  `x = sqrt(x^2+y^2+z^2) sin(theta) cos(phi)` and `y = sqrt(x^2+y^2+z^2) sin(theta) sin(phi)`?
 
If not, then I'm worried about incompatibilities between these definitions and all the other Kerr-Schild definitions everywhere else in the code.  (i.e. the horizon "radius" x^2+y^2+z^2 as a function of angles will be different here than everywhere else in the code).

If so, then I think these functions can be simplified.

Please add something in the documentation about this.

Reply:
Ok, I think I have figured out theta and phi coordinates.  We need to be very careful with the documentation and usage because I don't want to mix them up.  I think (please verify this in case I messed up) that we now have **three** sets of angular coordinates that we use for a Kerr-Schild metric in SpEC and SpECTRE.  Given the usual Kerr-Schild x,y,z coordinates as defined in MTW exercise 33.8 and elsewhere, here are our definitions (x,y,z mean the same thing in all these definitions):

1.  The usual spherical coordinates r, theta, phi defined from x,y,z in the usual way.
     `x = r sin(theta) cos(phi)`,   `y = r sin(theta) sin(phi)`,  `z = r cos(theta)`

     These are r,theta,phi that we use, e.g. when finding a horizon r as a function of theta,phi in apparent horizon finding (in SpEC and SpECTRE), and in SpEC waveform extraction (expanded in Ylms as a function of this theta,phi).  The various Kerr Horizon tests in Test_StrahlkorperGr.cpp use these angular coordinates. Notice that the horizon of a Kerr BH is *not* at a constant value of r.

1.  A spheroidal coordinate system r_BL, theta', phi
     `x = sqrt(r_BL^2 + a^2) sin(theta') cos(phi)`,  `y = sqrt(r_BL^2 + a^2) sin(theta') sin(phi)`,  `z = r_BL cos(theta')`

     The phi coordinate is the same as in the previous definition, but `r_BL != r` and `theta' != theta`.  Here `r_BL` is the Boyer-Lindquist radial coordinate, so the horizon of a Kerr black hole is at a constant value of `r_BL`.  I think these are the coordinates used in the "Kerr conforming" map in SpEC initial data.  At least this is something that appears in SpEC documentation but I haven't checked the map itself.

1. The coordinates defined in this PR: r_BL, theta', phi':

   `x = (r_BL cos(phi')+a sin(phi')) sin(theta')`,  `y = (r_BL sin(phi')-a cos(phi'))sin(theta')`,  `z = r_BL cos(theta')`
    The r_BL and theta' coordinates are the same as in the previous definition, but `phi' != phi`.   I think that `phi' = phi + arctan(a/r_BL)` but I haven't checked this.

I suggest that the documentation lay out these three definitions of angular coordinates, explicitly state that the three definitions are different, and (if possible) give theta, phi different names here than elsewhere to emphasize that they are different coordinates.